torch/_inductor/comms.py

```diff
@@ -323,40 +455,100 @@ def is_groupable(candidate):
                         candidate = _prev[candidate]
                         continue
                     else:
                         msg = (
                             f"data dependency {data_dep}(dep_names:{list(data_deps.keys())})"
-                            f"\n candidate:{candidate.get_name()}(os:{[candidate.get_buffer_names()]})"
-                            f"dep on {_group_names(group_head, group_tail)}"
-                            f"\n non_group_reason:{grp_reason}"
+                            f"\n candidate:{candidate.get_name()}(outs:{[candidate.get_buffer_names()]})"
+                            f"dep on {_group_names(gns)}"
+                            f"\n non_group_reason:{grouping_reason}"
                         )
-                        reorder_info.limiting_factor = msg
+                        info.limiting_factor = msg
                         break
 
-                delta_memory_candidate = (
-                    _curr_memory[candidate] - _curr_memory[_prev[candidate]]  # type: ignore[index]
+                candidate_allocfree: SNodeMemory = snodes_allocfree[candidate]
+                candidate_delta_mem: int = (
+                    candidate_allocfree.size_alloc - candidate_allocfree.size_free
                 )
+                gn_to_bufs_last_use: dict[
+                    BaseSchedulerNode, list[Union[FreeableInputBuffer, Any]]
+                ] = defaultdict(list)
+                for (
+                    buf,
+                    snode_last_use,
+                ) in buf_to_snode_last_use.items():
+                    succ_nodes = buf.mpi_buffer.succ_nodes
+                    if candidate not in succ_nodes:
+                        continue
 
-                if group_peak_memory - delta_memory_candidate > peak_memory:
-                    reorder_info.limiting_factor = "peak memory"
-                    break
+                    if not any(gn == snode_last_use for gn in gns):
+                        continue
 
-                reorder_info.moves += 1
+                    # candidate and one of group nodes are successors of the same buffer
+                    # and last use of the buffer happen in group nodes.
+                    # This last use deallocates it.
+                    # If we swap [candidate [group]] to [[group] candidate],
+                    # candidate becomes the last use
+                    # and deallocated this buffer instead of group node.
+                    # we need to update size_free accordingly to group_node and candidate,
+                    # and recalculate post_alloc, post_free for them.
+                    gn_to_bufs_last_use[snode_last_use].append(buf)
+
+                # Caching calculations of memory for group nodes and candidate,
+                # to apply without recalculation after swap.
+                _post_alloc_update: dict[BaseSchedulerNode, int] = {}
+
+                potential_peak: int = 0
+                if gn_to_bufs_last_use:
```

Comment:
might help readability to extract a helper or two here.  calculate_peak(), update_peak_... ? idk, maybe its not any different

Reply:
+1 - i would find it a little bit easier to read with some of the helpers abstracted out